dev/system-test/firestore.ts

```diff
@@ -2544,6 +2544,444 @@ describe('Query class', () => {
+      results = await collection
+        .where('key', '>=', 'a')
+        .where('sort', '<=', 2)
+        .orderBy('v', 'desc')
+        .limitToLast(2)
+        .get();
+      expectDocs(results, 'doc4', 'doc1');
+    });
+
+    it('can use on special values', async () => {
+      const collection = await testCollectionWithDocs({
+        doc1: {key: 'a', sort: 0, v: 0},
+        doc2: {key: 'b', sort: NaN, v: 1},
+        doc3: {key: 'c', sort: null, v: 3},
+        doc4: {key: 'd', v: 0},
+        doc5: {key: 'e', sort: 1},
+        doc6: {key: 'f', sort: 1, v: 1},
+      });
+
+      let results = await collection
+        .where('key', '!=', 'a')
+        .where('sort', '<=', 2)
+        .get();
+      expectDocs(results, 'doc5', 'doc6');
+
+      results = await collection
+        .where('key', '!=', 'a')
+        .where('sort', '<=', 2)
+        .where('v', '<=', 1)
+        .get();
+      expectDocs(results, 'doc6');
+    });
+
+    it('can use with array membership', async () => {
+      const collection = await testCollectionWithDocs({
+        doc1: {key: 'a', sort: 0, v: [0]},
+        doc2: {key: 'b', sort: 1, v: [0, 1, 3]},
+        doc3: {key: 'c', sort: 1, v: []},
+        doc4: {key: 'd', sort: 2, v: [1]},
+        doc5: {key: 'e', sort: 3, v: [2, 4]},
+        doc6: {key: 'f', sort: 4, v: [NaN]},
+        doc7: {key: 'g', sort: 4, v: [null]},
+      });
+
+      let results = await collection
+        .where('key', '!=', 'a')
+        .where('sort', '>=', 1)
+        .where('v', 'array-contains', 0)
+        .get();
+      expectDocs(results, 'doc2');
+
+      results = await collection
+        .where('key', '!=', 'a')
+        .where('sort', '>=', 1)
+        .where('v', 'array-contains-any', [0, 1])
+        .get();
+      expectDocs(results, 'doc2', 'doc4');
+    });
+
+    // Use cursor in following test cases to add implicit order by fields in the sdk and compare the
+    // result with the query fields normalized in the server.
+    it('can use with nested field', async () => {
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const testData = (n?: number): any => {
+        n = n || 1;
+        return {
+          name: 'room ' + n,
+          metadata: {
+            createdAt: n,
+          },
+          field: 'field ' + n,
+          'field.dot': n,
+          'field\\slash': n,
+        };
+      };
+
+      const collection = await testCollectionWithDocs({
+        doc1: testData(400),
+        doc2: testData(200),
+        doc3: testData(100),
+        doc4: testData(300),
+      });
+
+      // ordered by: name asc, metadata.createdAt asc, __name__  asc
+      let query = collection
+        .where('metadata.createdAt', '<=', 500)
+        .where('metadata.createdAt', '>', 100)
+        .where('name', '!=', 'room 200')
+        .orderBy('name');
+      let docSnap = await collection.doc('doc4').get();
+      let query_sdk_normalized = query.startAt(docSnap);
```

Comment:
Oops. Shouldn't mix the use of snake case (e.g. `query_sdk_normalized`) and camel case (e.g. `docSnap`).

Reply:
also, we can probably use a better name than "sdk normalized" . such as: queryWithCursor.

The comment above (line 2669) makes it clear that we're adding cursors to trigger the implicit orderBy logic.